simulator.py

```diff
@@ -0,0 +1,97 @@
+"""
+Create a simulator to calculate pnl
+"""
+
+import copy
+import uuid
+from thousandaire.dataset import Data
+
+DATASET = 'price_dataset'
+
+def decode_data(data):
+    """
+    Decode while in multi-process.
+
+    In multi-process, simulator will get the data reference
+    which alpha need. To prevent polluteing the origin data, we need to
+    copy a local version.
+    Since 'target_data' may also in 'others', that we should deepcopy the data
+    separately not simply using copy.deepcopy(data).
+    """
+    return {
+        'workdays' : copy.deepcopy(data['workdays']),
+        'target_data' : copy.deepcopy(data['target_data']),
+        'others' : copy.deepcopy(data['others'])}
+
+class Simulator:
+    """
+    Simulate the alpha.
+    """
+    def __init__(self, settings, data, cost_function):
+        self.cost_function = cost_function
+        self.data = decode_data(data)
+        self.settings = settings
+        self.__result = Data(
+            'result', ['pnl', 'cost', 'positions_raw', 'positions_np'])
+        self.__portfolio = list()
+        self.__key = uuid.uuid4()
+        self.initialize_data()
+
+    def next_date(self):
+        """
+        Move the simulating date.
+        """
+        for dataset in self.data['others'].values():
+            dataset.next_date(self.__key)
+        self.data['target_data'].next_date(self.__key)
+        self.data['workdays'].next_date(auth_key=self.__key)
+
+    def produce_pnl(self):
+        """
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        Profit and cost will return by user-specified cost_function.
+        """
+        self.cost_function.reset()
+        self.cost_function(self.__portfolio, self.data['target_data'])
+        profit, cost = self.cost_function.result()

```

Comment:
Can we just let it be `profit, cost = self.cost_function(self.__portfolio, self.data['target_data'])`?